src/mixtures/mixturemodel.jl

```diff
@@ -367,15 +367,31 @@ function _mixlogpdf!(r::AbstractArray, d::AbstractMixtureModel, x)
     return r
 end
 
-function _mixgradlogpdf1(d::AbstractMixtureModel, x)
-    pdfdx = pdf(d, x) 
-    glp = !iszero(pdfdx) ? sum(pi * pdf(d.components[i], x) .* gradlogpdf(d.components[i], x) for (i, pi) in enumerate(probs(d)) if (!iszero(pi) && !iszero(pdf(d.components[i], x)))) / pdfdx : pdfdx
-    return glp
-end
-
 pdf(d::UnivariateMixture, x::Real) = _mixpdf1(d, x)
 logpdf(d::UnivariateMixture, x::Real) = _mixlogpdf1(d, x)
-gradlogpdf(d::UnivariateMixture, x::Real) = _mixgradlogpdf1(d, x)
+
+function gradlogpdf(d::UnivariateMixture, x::Real)
+    cp = components(d)
+    pr = probs(d)
+    pdfx1 = pdf(cp[1], x)
+    pdfx = pr[1] * pdfx1
+    _glp = pdfx * gradlogpdf(cp[1], x)
+    glp = (!iszero(pr[1])) && (!iszero(pdfx)) ? _glp : zero(_glp)
+    @inbounds for i in Iterators.drop(eachindex(pr, cp), 1)
```

Comment:
I am not sure how to avoid duplicating the operations. I have to have `glp` out of the loop and initiated with correct type. I will think more about it and about how to use `iterate`. I actually like `Iterators.drop`, but if there is a way to avoid duplicating the code, that's better.

Reply:
What I meant is: You retrieve the first elements (and the state) with `iterate(pr)` and `iterate(cp)` outside of the loop (and then you already have the elements and don't need to call `cp[1]` etc. multiple times), and then in the for loop (maybe rather a while loop then) you would call `iterate(pr, pr_state)` etc. to retrieve the next element and the new state (and again you don't need to call `pr[i]` etc. multiple times).

Reply:
Ok. I see. Is there any performance advantage with that or is it more a matter of taste? I will give it a try anyway. Thanks.

Reply:
I don't expect a performance difference (the indexing shouldn't matter compared with the actual computational cost) but IMO there are a few other advantages: It does not assume 1-based indexing, it avoids `Iterators.drop` etc, and the iteration/indices of both arrays are kept completely separate.

Reply:
Oh, of course, I should have used `first()` instead of `[1]`. But, yeah, `iterate` each one does not assume they have a common indexing. Good.

Reply:
This is the best I came up with.

I don't like much `while true` loops, but that is what `sum()` essentially does (it boils down to https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L53) and it avoids some extra calsl to `iterate()`.